### PR TITLE
PREAPPS-4827: Implement GetCustomMetadata and SetCustomMetadata

### PR DESCRIPTION
**Pull request by lavegupta**

### CODE REVIEW

File: src/normalize/entities.ts [hunk 1]
```diff
@@ -415,11 +415,21 @@ const ContactAttributes = new Entity({
 	image: ImageFields
 });
 
+const CustomMetaAttributes = new Entity({
+	chatConversationId: 'chatConversationId'
```
Comment (silentsakky) [inline src/normalize/entities.ts:419]: zimlet developers will need flexibility to define their own meta tags, so hardcoding here will not help them

File: src/normalize/entities.ts [hunk 2]
```diff
@@ -415,11 +415,21 @@ const ContactAttributes = new Entity({
...
+	section: 'section',
+	_attrs: ['attributes', CustomMetaAttributes]
+});
+
 const contactFields = {
 	d: 'date',
 	l: 'folderId',
 	rev: 'revision',
 	sf: 'sortField',
+	meta: ['meta', CustomMetadata],
```
Comment (silentsakky) [inline src/normalize/entities.ts:432]: this is not required

File: src/normalize/entities.ts [hunk 3]
```diff
@@ -415,11 +415,21 @@ const ContactAttributes = new Entity({
 	image: ImageFields
 });
 
+const CustomMetaAttributes = new Entity({
+	chatConversationId: 'chatConversationId'
+});
+
+const CustomMetadata = new Entity({
+	section: 'section',
```
Comment (silentsakky) [inline src/normalize/entities.ts:423]: if key and value are same then we don't need to specify it here

File: src/schema/schema.graphql [hunk 4]
```diff
@@ -1388,6 +1402,7 @@ type Contact {
 	memberOf: String
 	attributes: ContactAttributes
 	members: [ContactListMember]
+	meta: [CustomMetadataMeta]
```
Comment (silentsakky) [inline src/schema/schema.graphql:1405]: we can remove this

File: src/schema/schema.ts [hunk 5]
```diff
@@ -410,6 +413,18 @@ export function createZimbraSchema(
 				revokeRights: (_, variables) =>
 					client.revokeRights(variables.input as RevokeRightsInput),
 				revokeTrustedDevice: client.revokeTrustedDevice,
+				setCustomMetadata: (_: any, variables: any) => client
+						.jsonRequest({
```
Comment (silentsakky) [inline src/schema/schema.ts:417]: we don't do jsonRequest from here directly, it needs to be moved to batch-client

Comment (lavegupta) [reply src/schema/schema.ts]: Done

File: src/schema/schema.graphql (hunk 4 above)
Comment (lavegupta) [reply src/schema/schema.graphql]: Done

File: src/normalize/entities.ts (hunk 2 above)
Comment (lavegupta) [reply src/normalize/entities.ts]: Done

File: src/normalize/entities.ts (hunk 1 above)
Comment (lavegupta) [reply src/normalize/entities.ts]: Done

File: src/utils/normalize-attrs-custommetadata.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export function normalizeCustomMetaDataAttrs(data: any) {
+	let attrs: any = [];
+
+	Object.keys(data._attrs).forEach(
+		key =>
+			typeof data._attrs[key] === 'string' &&
+			attrs.push({ key, value: data._attrs[key] }) &&
+			delete data._attrs[key]
```
Comment (silentsakky) [inline src/utils/normalize-attrs-custommetadata.ts:28]: why are we deleting keys here?

Comment (lavegupta) [reply src/utils/normalize-attrs-custommetadata.ts]: Removed

**Pull request merged**